### PR TITLE
1.0: Unify promise-based API; improve software architecture

### PR DESCRIPTION
**Pull request by namoscato**
Resolves https://github.com/vue-gapi/vue-gapi/issues/105

## Checklist

- [x] Abstract session storage implementation
- [x] Update documentation
- [x] Update / expand test coverage

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
+export default {
+  install: (Vue, clientConfig) => {
+    const gapiClientProvider = new GapiClientProvider(
+      getObjectCopy(clientConfig)
+    )
 
-// Install by default if using the script tag
-if (typeof window !== 'undefined' && window.Vue) {
-  window.Vue.use(plugin)
+    Vue.prototype.$gapi = new GoogleAuthService(gapiClientProvider)
```
Comment (namoscato) [inline src/index.js:11]: @cedpoilly / others, I would love some feedback on this approach which exposes the `GoogleAuthService` instance directly and introduces a `GapiClientProvider` dependency (open to suggestions on naming).

This will most noticeably affect the implementation of `GoogleAuthService` methods which interact with `GapiClientProvider`'s promise-based methods in favor of referencing an `authInstance` property which was set externally. I've updated `login` / `logout` as an example of what this change would look like.

This will likely be a breaking change and could give us a reason to move to Vue v3.x and/or improve APIs while we're at it. Let me know what you think.

File: src/VueGapi/GoogleAuthService.js [hunk 2]
```diff
@@ -254,20 +246,11 @@ export default class GoogleAuthService {
...
-        (error) => {
-          rej(error)
-        }
-      )
-    }).then(...thenArgsFromCallbacks(onResolve, onReject))
+  logout() {
+    return this.clientProvider
+      .getAuthInstance()
+      .then((authInstance) => authInstance.signOut())
+      .then(() => this._clearStorage())
```
Comment (deCodeIt) [inline src/VueGapi/GoogleAuthService.js:253]: I prefer this approach as it's compact and clear. Rather than having `onResolve` and `onReject` methods as callbacks, letting the user use async-await on `logout()` or treat logout as a promise is a better approach.

File: src/VueGapi/GapiClientProvider.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            .init(this.clientConfig)
+            .then(() => resolve((this.gapi = gapi)), reject)
+        })
+      })
+    })
+  }
+
+  getAuthInstance() {
+    return this.getClient().then((gapi) => gapi.auth2.getAuthInstance())
+  }
```
Comment (deCodeIt) [inline src/VueGapi/GapiClientProvider.js:36]: If we can also expose the `authInstance` to caller, something like `this.$gapi.getAuthInstance()` then that will help the users write their own **additional** functionality.

File: src/index.js (hunk 1 above)
Comment (deCodeIt) [reply src/index.js:11]: I prefer a promise based approach on the function rather than onResolve, onReject callbacks. As your approach is a promise based approach, the design changes look good to me ( LGTM! ). Also it's better if we also expose the `authInstancce` to the user as `this.$gapi.getAuthInstance()` as the user will be able to implement their own functionality extending the library. Readme should mention to use the library's login / logout as we do use localStorage for cache mechanism.

**Review by deCodeIt — COMMENTED** [review]
Tl;DR: The design change LGTM!

File: src/VueGapi/GapiClientProvider.js (hunk 3 above)
Comment (namoscato) [reply src/VueGapi/GapiClientProvider.js]: Good call. 3ef77dceb658def3d7f13ce83b73b62ddaf429d3 exposes `getAuthInstance` _and_ `getCurrentUser`.

**Review by cedpoilly — APPROVED** [review]
